nexus/types/src/internal_api/views.rs

```diff
@@ -583,40 +599,109 @@ impl UpdateStatus {
                 )
             })
             .collect();
-        UpdateStatus { zones }
+        let baseboard_ids: Vec<_> = inventory.sps.keys().cloned().collect();
+
+        // Find all SP versions and git commits via cabooses
+        let mut sps: BTreeMap<BaseboardId, SpStatus> = baseboard_ids
+            .into_iter()
+            .map(|baseboard_id| {
+                let slot0_version = inventory
+                    .caboose_for(CabooseWhich::SpSlot0, &baseboard_id)
+                    .map_or(TufRepoVersion::Unknown, |c| {
+                        Self::caboose_to_version(old, new, &c.caboose)
+                    });
+                let slot1_version = inventory
+                    .caboose_for(CabooseWhich::SpSlot1, &baseboard_id)
+                    .map_or(TufRepoVersion::Unknown, |c| {
+                        Self::caboose_to_version(old, new, &c.caboose)
+                    });
+                (
+                    (*baseboard_id).clone(),
+                    SpStatus { sled_id: None, slot0_version, slot1_version },
+                )
+            })
+            .collect();
+
+        // Fill in the sled_id for the sp if known
+        for sa in inventory.sled_agents.iter() {
+            if let Some(baseboard_id) = &sa.baseboard_id {
+                if let Some(sp) = sps.get_mut(baseboard_id) {
+                    sp.sled_id = Some(sa.sled_id);
+                }
+            }
+        }
+
+        let sps = sps.into_iter().map(|(k, v)| (k.to_string(), v)).collect();
+
+        UpdateStatus { zones, sps }
+    }
+
+    fn caboose_to_version(
+        old: &TargetReleaseDescription,
+        new: &TargetReleaseDescription,
+        caboose: &Caboose,
+    ) -> TufRepoVersion {
+        let matching_caboose = |a: &TufArtifactMeta| {
+            caboose.board == a.id.name
+                && matches!(
+                    a.id.kind.to_known(),
+                    Some(
+                        KnownArtifactKind::GimletSp
+                            | KnownArtifactKind::PscSp
+                            | KnownArtifactKind::SwitchSp
+                    )
+                )
+                && caboose.version == a.id.version.to_string()
+        };
+        if let Some(old) = old.tuf_repo() {
+            if old.artifacts.iter().any(matching_caboose) {
+                return TufRepoVersion::Version(
+                    old.repo.system_version.clone(),
+                );
+            }
+        }
+        if let Some(new) = new.tuf_repo() {
```

Comment:
I think it's possible (albeit unlikely) SP versions don't change between releases - maybe put this one before checking `old`, so we report the newer if it matches both?

Reply:
Good idea! One other thing, is that testing this with omicron-dev does not seem to be doing what I expect. I also changed to parse the `ArtifactVersion` in case `to_string` was not symmetric with `parse`. Didn't make a difference. I think this is likely just an artifact of my test environment, but can't be sure without testing on a4x2 or a racklette.